core/sourcehook/generate/sourcehook.hxx

```diff
@@ -566,6 +570,730 @@ namespace SourceHook
+		*
+		*/
+		template<typename Result, typename... Args>
+		struct BaseMethodInvoker
+		{
+		public:
+			typedef Result (EmptyClass::*EmptyDelegate)(Args...);
+		};
+
+		template<typename IDelegate, typename... Args>
+		struct VoidMethodInvoker
+		{
+		public:
+			typedef std::bool_constant<false> has_return;
+			typedef BaseMethodInvoker<void, Args...> base;
+
+			static void Invoke(IDelegate* delegate, void* result, Args... args)
+			{
+				//	Do not touch return type: It's void!
+				delegate->Call(args...);
+			}
+
+			static void Original(EmptyClass* self, typename base::EmptyDelegate mfp, void* result, Args... args)
+			{
+				//	Do not touch return type: It's void!
+				(self->*mfp)(args...);
+			}
+
+			template<typename Self, typename Mfp>
+			static void OriginalRaised( void (*Invoker)(Self, Mfp, Args...), Self self, Mfp mfp, void* result, Args... args )
+			{
+				//	Do not touch return type: It's void!
+				Invoker(self, mfp, args...);
+			}
+
+			static void Dereference(const void* arg)
+			{
+				return;
+			}
+		};
+
+		template<typename IDelegate, typename Result, typename... Args>
+		struct ReturningMethodInvoker
+		{
+		public:
+			typedef std::bool_constant<true> has_return;
+			typedef BaseMethodInvoker<Result, Args...> base;
+
+
+			static void Invoke(IDelegate* delegate, Result* result, Args... args)
+			{
+				*result = delegate->Call(args...);
+			}
+
+			static void Original(EmptyClass* self, typename base::EmptyDelegate mfp, Result* result, Args... args)
+			{
+				*result = (self->*mfp)(args...);
+			}
+
+			template<typename Self, typename Mfp>
+			static void OriginalRaised( Result (*Invoker)(Self, Mfp, Args...), Self self, Mfp mfp, Result* result, Args... args )
+			{
+				*result = Invoker(self, mfp, args...);
+			}
+
+			static Result Dereference(const Result* arg)
+			{
+				return *arg;
+			}
+		};
+
+		/**
+		 * @brief defines the interface from the hook manager -> user code
+		 *
+		 * This is invoked from the hook manager to call SourceHook delegates once
+		 * the hook manager has lowered arguments/etc into the receiving delegate types.
+		 */
+		template<ProtoInfo::CallConvention Convention, typename Result, typename... Args>
+		struct HookHandlerImpl
+		{
+		public:
+			/**
+			 * @brief The delegate type that SourceHook will invoke (user code)
+			 */
+			typedef typename fastdelegate::FastDelegate<Result, Args...> Delegate;
+
+			struct IMyDelegate : ::SourceHook::ISHDelegate
+			{
+			public:
+				virtual Result Call(Args... args) = 0;
+			};
+
+			struct CMyDelegateImpl : IMyDelegate
+			{
+			public:
+				Delegate _delegate;
+
+				CMyDelegateImpl(Delegate deleg) : _delegate(deleg)
+				{}
+				virtual~CMyDelegateImpl() {}
+				Result Call(Args... args) { return _delegate(args...); }
+				void DeleteThis() { delete this; }
+				bool IsEqual(ISHDelegate *pOtherDeleg) { return _delegate == static_cast<CMyDelegateImpl *>(pOtherDeleg)->_delegate; }
```

Comment:
This looks very suspicious
`static_cast<CMyDelegateImpl *>`
The other delegate is not guaranteed to be a `CMyDelegateImpl`, this might not be the case here and now. But from an API standpoint `IsEqual` can be called with any classes implementing `SHDelegate`, one notable such class could be dynamically generated hooks.

I suggest swapping this out for a `dynamic_cast` or return false all the time.

Reply:
That’s actually what the [macros do](https://github.com/alliedmodders/metamod-source/blob/8abbfbb8106bfe79b0096d981bbc337e58a44052/core/sourcehook/generate/sourcehook.hxx#L1780).

Would there be any behavior change using dynamic_cast if any plugins running under Metamod were compiled without RTTI?

Reply:
Relaying my thoughts from discord.

After carefully reading SH, `IsEqual` is currently only invoked when a hook is removed. Meaning whatever delegate is passed will be a `CMyDelegate`. So you can actually ensure the class type the function will be invoked with... Until we start accounting dynamic hooks, their `ISHDelegate` implem can be empty, and if those are ever removed. Then `ISHDelegate` will not be a `CMyDelegate`.

Ultimately since `_delegate` is only `+4` or `+8` from the this ptr, IF `pOtherDeleg` isn't a `CMyDelegate` it won't fall too far away from the address and most likely still be in a read allowed space. Buuuut there's still a risk, and it might mislead future contributors about `bool IsEqual(ISHDelegate *pOtherDeleg)` like it did for you.

Then there's the issue you raise with RTTI and dynamic cast. And this currently seems like an unsolvable problem. Not without either modifying the virtual function definition to provide more details about the delegate, or make a new one.

Reply:
IsEqual() should only be called on ISHDelegates within the same plugin (eg binary) *and* the same HookManPubFunc, which means there's no way it'll be called on ISHDelegates that were macro-expanded/compiled in different places.

We have to wrap our heads around this every time it comes up, so it's probably worth a comment in the code somewhere, and maybe a test mixing hook types (eg a macro hook and a template hook on the same vtable entry) proving that there are no cross-calls.

Reply:
Oh yeah true, I hadn't realised `HookManPubFunc` guards against that issues.

As for the test I can do that in the revised sourcehook test suite I've added in my own PR.

Reply:
Cool! I'll mark this as resolved then. Sticking with static casts for now